openhands/llm/llm.py

```diff
@@ -207,14 +211,21 @@ def custom_completion_wait(retry_state):
                 max=max_wait_time,
             )
 
+            # Check for shutdown before returning the wait time
+            if should_exit():
+                raise UserCancelledError('Shutdown requested during retry by user.')
+
             # Call the exponential wait function with retry_state to get the actual wait time
             return exponential_wait(retry_state)
 
         @retry(
             after=attempt_on_error,
             stop=stop_after_attempt(self.config.num_retries),
             reraise=True,
-            retry=retry_if_exception_type(self.retry_exceptions),
+            retry=(
+                retry_if_exception_type(self.retry_exceptions)
+                & retry_if_not_exception_type(UserCancelledError)
```

Comment:
just a bit curious here, `self.retry_exceptions` is [this](https://github.com/All-Hands-AI/OpenHands/pull/3968/files#diff-b46ec9dd4b319d75809520e34b2559fd2ce46ca4895709eae9312aa04a8fa9a4R89) if I see this right? it doesn't have UserCancelledError - of course we don't want to. So the second part will never be hit

Reply:
The 2nd part will be hit from the `before_sleep` which does end the retry loop.